File: python/inkbox_mail/resources/messages.py
"""
inkbox_mail/resources/messages.py

Message operations: list (auto-paginated), get, send, flag updates, delete.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, AsyncIterator
from uuid import UUID

from inkbox_mail.types import Message, MessageDetail

if TYPE_CHECKING:
    from inkbox_mail._http import HttpTransport

_DEFAULT_PAGE_SIZE = 50


class MessagesResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    async def list(
        self,
        mailbox_id: UUID | str,
        *,
        page_size: int = _DEFAULT_PAGE_SIZE,
    ) -> AsyncIterator[Message]:
        """Async iterator over all messages in a mailbox, newest first.

        Pagination is handled automatically — just iterate.

        Args:
            mailbox_id: UUID of the mailbox.
            page_size: Number of messages fetched per API call (1–100).

        Example::

            async for msg in client.messages.list(mailbox_id):
                print(msg.subject, msg.from_address)
        """
        return self._paginate(mailbox_id, page_size=page_size)

    async def _paginate(
        self,
        mailbox_id: UUID | str,
        *,
        page_size: int,
    ) -> AsyncIterator[Message]:
        cursor: str | None = None
        while True:
            page = await self._http.get(
                f"/mailboxes/{mailbox_id}/messages",
                params={"limit": page_size, "cursor": cursor},
            )
            for item in page["items"]:
                yield Message._from_dict(item)
            if not page["has_more"]:
                break
            cursor = page["next_cursor"]

    async def get(self, mailbox_id: UUID | str, message_id: UUID | str) -> MessageDetail:
        """Get a message with full body content.

        Args:
            mailbox_id: UUID of the owning mailbox.
            message_id: UUID of the message.

        Returns:
            Full message including ``body_text`` and ``body_html``.
        """
        data = await self._http.get(f"/mailboxes/{mailbox_id}/messages/{message_id}")
        return MessageDetail._from_dict(data)

    async def send(
        self,
        mailbox_id: UUID | str,
        *,
        to: list[str],
        subject: str,
        body_text: str | None = None,
        body_html: str | None = None,
        cc: list[str] | None = None,
        bcc: list[str] | None = None,
        in_reply_to_message_id: str | None = None,
    ) -> Message:
        """Send an email from a mailbox.

        Args:
            mailbox_id: UUID of the sending mailbox.
            to: Primary recipient addresses (at least one required).
            subject: Email subject line.
            body_text: Plain-text body.
            body_html: HTML body.
            cc: Carbon-copy recipients.
            bcc: Blind carbon-copy recipients.
            in_reply_to_message_id: RFC 5322 Message-ID of the message being
                replied to. Threads the reply automatically.

        Returns:
            The sent message metadata.
        """
        recipients: dict[str, Any] = {"to": to}
        if cc:
            recipients["cc"] = cc
        if bcc:
            recipients["bcc"] = bcc

        body: dict[str, Any] = {"recipients": recipients, "subject": subject}
        if body_text is not None:
            body["body_text"] = body_text
        if body_html is not None:
            body["body_html"] = body_html
        if in_reply_to_message_id is not None:
            body["in_reply_to_message_id"] = in_reply_to_message_id

        data = await self._http.post(f"/mailboxes/{mailbox_id}/messages", json=body)
        return Message._from_dict(data)

    async def update_flags(
        self,
        mailbox_id: UUID | str,
        message_id: UUID | str,
        *,
        is_read: bool | None = None,
        is_starred: bool | None = None,
    ) -> Message:
        """Update read/starred flags on a message.

        Pass only the flags you want to change; omitted flags are left as-is.
        """
        body: dict[str, Any] = {}
        if is_read is not None:
            body["is_read"] = is_read
        if is_starred is not None:
            body["is_starred"] = is_starred
        data = await self._http.patch(
            f"/mailboxes/{mailbox_id}/messages/{message_id}", json=body
        )
        return Message._from_dict(data)

    async def mark_read(self, mailbox_id: UUID | str, message_id: UUID | str) -> Message:
        """Mark a message as read."""
        return await self.update_flags(mailbox_id, message_id, is_read=True)

    async def mark_unread(self, mailbox_id: UUID | str, message_id: UUID | str) -> Message:
        """Mark a message as unread."""
        return await self.update_flags(mailbox_id, message_id, is_read=False)

    async def star(self, mailbox_id: UUID | str, message_id: UUID | str) -> Message:
        """Star a message."""
        return await self.update_flags(mailbox_id, message_id, is_starred=True)

    async def unstar(self, mailbox_id: UUID | str, message_id: UUID | str) -> Message:
        """Unstar a message."""
        return await self.update_flags(mailbox_id, message_id, is_starred=False)

    async def delete(self, mailbox_id: UUID | str, message_id: UUID | str) -> None:
        """Delete a message."""
        await self._http.delete(f"/mailboxes/{mailbox_id}/messages/{message_id}")

File: typescript/src/resources/messages.ts
/**
 * inkbox-mail/resources/messages.ts
 *
 * Message operations: list (auto-paginated), get, send, flag updates, delete.
 */

import { HttpTransport } from "../_http.js";
import {
  Message,
  MessageDetail,
  RawCursorPage,
  RawMessage,
  parseMessage,
  parseMessageDetail,
} from "../types.js";

const DEFAULT_PAGE_SIZE = 50;

export class MessagesResource {
  constructor(private readonly http: HttpTransport) {}

  /**
   * Async iterator over all messages in a mailbox, newest first.
   *
   * Pagination is handled automatically — just iterate.
   *
   * @example
   * ```ts
   * for await (const msg of client.messages.list(mailboxId)) {
   *   console.log(msg.subject, msg.fromAddress);
   * }
   * ```
   */
  async *list(
    mailboxId: string,
    options?: { pageSize?: number },
  ): AsyncGenerator<Message> {
    const limit = options?.pageSize ?? DEFAULT_PAGE_SIZE;
    let cursor: string | undefined;

    while (true) {
      const page = await this.http.get<RawCursorPage<RawMessage>>(
        `/mailboxes/${mailboxId}/messages`,
        { limit, cursor },
      );
      for (const item of page.items) {
        yield parseMessage(item);
      }
      if (!page.has_more) break;
      cursor = page.next_cursor ?? undefined;
    }
  }

  /**
   * Get a message with full body content.
   *
   * @param mailboxId - UUID of the owning mailbox.
   * @param messageId - UUID of the message.
   */
  async get(mailboxId: string, messageId: string): Promise<MessageDetail> {
    const data = await this.http.get<RawMessage>(
      `/mailboxes/${mailboxId}/messages/${messageId}`,
    );
    return parseMessageDetail(data);
  }

  /**
   * Send an email from a mailbox.
   *
   * @param mailboxId - UUID of the sending mailbox.
   * @param options.to - Primary recipient addresses (at least one required).
   * @param options.subject - Email subject line.
   * @param options.bodyText - Plain-text body.
   * @param options.bodyHtml - HTML body.
   * @param options.cc - Carbon-copy recipients.
   * @param options.bcc - Blind carbon-copy recipients.
   * @param options.inReplyToMessageId - RFC 5322 Message-ID of the message being
   *   replied to. Threads the reply automatically.
   * @param options.attachments - Optional file attachments. Each entry must have
   *   `filename`, `contentType` (MIME type), and `contentBase64` (base64-encoded
   *   file content). Max total size: 25 MB. Blocked: `.exe`, `.bat`, `.scr`.
   */
  async send(
    mailboxId: string,
    options: {
      to: string[];
      subject: string;
      bodyText?: string;
      bodyHtml?: string;
      cc?: string[];
      bcc?: string[];
      inReplyToMessageId?: string;
      attachments?: Array<{
        filename: string;
        contentType: string;
        contentBase64: string;
      }>;
    },
  ): Promise<Message> {
    const recipients: Record<string, unknown> = { to: options.to };
    if (options.cc) recipients["cc"] = options.cc;
    if (options.bcc) recipients["bcc"] = options.bcc;

    const body: Record<string, unknown> = {
      recipients,
      subject: options.subject,
    };
    if (options.bodyText !== undefined) body["body_text"] = options.bodyText;
    if (options.bodyHtml !== undefined) body["body_html"] = options.bodyHtml;
    if (options.inReplyToMessageId !== undefined) {
      body["in_reply_to_message_id"] = options.inReplyToMessageId;
    }
    if (options.attachments !== undefined) {
      body["attachments"] = options.attachments.map((a) => ({
        filename: a.filename,
        content_type: a.contentType,
        content_base64: a.contentBase64,
      }));
    }

    const data = await this.http.post<RawMessage>(
      `/mailboxes/${mailboxId}/messages`,
      body,
    );
    return parseMessage(data);
  }

  /**
   * Update read/starred flags on a message.
   *
   * Pass only the flags you want to change; omitted flags are left as-is.
   */
  async updateFlags(
    mailboxId: string,
    messageId: string,
    flags: { isRead?: boolean; isStarred?: boolean },
  ): Promise<Message> {
    const body: Record<string, boolean> = {};
    if (flags.isRead !== undefined) body["is_read"] = flags.isRead;
    if (flags.isStarred !== undefined) body["is_starred"] = flags.isStarred;

    const data = await this.http.patch<RawMessage>(
      `/mailboxes/${mailboxId}/messages/${messageId}`,
      body,
    );
    return parseMessage(data);
  }

  /** Mark a message as read. */
  async markRead(mailboxId: string, messageId: string): Promise<Message> {
    return this.updateFlags(mailboxId, messageId, { isRead: true });
  }

  /** Mark a message as unread. */
  async markUnread(mailboxId: string, messageId: string): Promise<Message> {
    return this.updateFlags(mailboxId, messageId, { isRead: false });
  }

  /** Star a message. */
  async star(mailboxId: string, messageId: string): Promise<Message> {
    return this.updateFlags(mailboxId, messageId, { isStarred: true });
  }

  /** Unstar a message. */
  async unstar(mailboxId: string, messageId: string): Promise<Message> {
    return this.updateFlags(mailboxId, messageId, { isStarred: false });
  }

  /** Delete a message. */
  async delete(mailboxId: string, messageId: string): Promise<void> {
    await this.http.delete(`/mailboxes/${mailboxId}/messages/${messageId}`);
  }
}
